easybuild/easyblocks/p/python.py

```diff
@@ -192,6 +192,30 @@ def patch_step(self, *args, **kwargs):
             # Ignore user site dir. -E ignores PYTHONNOUSERSITE, so we have to add -s
             apply_regex_substitutions('configure', [(r"(PYTHON_FOR_BUILD=.*-E)'", r"\1 -s'")])
 
+        # If we filter out LD_LIBRARY_PATH (not unusual when using rpath), ctypes is not able to dynamically load
+        # libraries installed with EasyBuild (see https://github.com/EESSI/software-layer/issues/192).
+        # ctypes is using GCC (and therefore LIBRARY_PATH) to figure out the full location but then only returns the
+        # soname, instead let's return the full path in this particular scenario
+        filtered_env_vars = build_option('filter_env_vars') or []
+        if 'LD_LIBRARY_PATH' in filtered_env_vars and 'LIBRARY_PATH' not in filtered_env_vars:
+            ctypes_util_py = os.path.join("Lib", "ctypes", "util.py")
+            orig_gcc_so_name = None
+            # Let's do this incrementally since we are going back in time
+            if LooseVersion(self.version) >= "3.9.1":
+                # From 3.9.1 to at least v3.12.4 there is only one match for this line
+                orig_gcc_so_name = "_get_soname(_findLib_gcc(name)) or _get_soname(_findLib_ld(name))"
+            if orig_gcc_so_name:
+                orig_gcc_so_name_regex = r'(\s*)' + re.escape(orig_gcc_so_name) + r'(\s*)'
+                updated_gcc_so_name = (
+                    "os.path.join(os.path.dirname(_findLib_gcc(name)), _get_soname(_findLib_gcc(name)))"
```

Comment:
```
_get_soname(_findLib_gcc(name))
```
can fail when `name` does not have the `SONAME` inside, and this the cause this to also fail. We will need another approach

Reply:
I went down the rabbit hole with you @ocaisa ...

I figured I'd just check all *.so files in our software prefix (local software stack) to get some idea of whether it is common for the SONAME field to not be set in the ELF header. Just from visual inspection, that seems to be the case about half of the time. I don't understand how they can just assume here https://github.com/python/cpython/blob/56c3815ba14c790d2e9a227b4ac0ead5e6b1e570/Lib/ctypes/util.py#L200 that the SONAME is always set. It's probably good practice to make sure your build system adds a -Wl,-soname  whenever you build a shared library, but it's clearly not something that is done in the vast majority of cases.

I was playing a bit with `find_library`, `_findLib_gcc` and `_get_soname` manually. What I was wondering was: why is just calling `_findLib_gcc`  not sufficient? Why would you take the dir from that, then call `_get_soname` to get the `SONAME` from the ELF header, and join those again? Note that this isn't even guaranteed to result in a valid _filename_. A shared libary could be called `/path/to/my/libfoo.so` and the `SONAME` in the ELF header could be `libbla.so`. It'd be strange, but it's not forbidden. More likely would be the scenario where the file would be called `libfoo.so.1.0` and the the `SONAME` in the header would be `libfoo.so.1`, indicating that this is compatible with an API interface version 1. If `libfoo.so.1.1` would still be compatible, the `SONAME` in the ELF header would then still be `libfoo.so.1`.

Anyway, I read more, see e.g. https://github.com/python/cpython/issues/54207#issuecomment-1093520249 and the actual docs on https://docs.python.org/3/library/ctypes.html 

> The purpose of the find_library() function is to locate a library in a way similar to what the compiler or runtime loader does (on platforms with several versions of a shared library the most recent should be loaded)

Well, that can't both be true. A compiler searches LIBRARY_PATH and a runtime loader searches LD_LIBRARY_PATH. So that documented behavior is a load of bullocks if you ask me. Nevertheless, the comment in https://github.com/python/cpython/issues/54207#issuecomment-1093520249 hits the nail on the head: if find_library is supposed to return the file that the runtime linker would load on e.g. a dlopen call, it should respect much more than LD_LIBRARY_PATH, namely also RPATH and RUNPATH . But there is one problem: RPATH  and RUNPATH are properties of an ELF executable or library. A python program doesn't have these properties... So, this is where python's `ctypes` can never help a python program 'behave' exactly the same as an ELF executable...

I guess the best thing we can do is indeed pretend that we are at 'compile time' again, and find what library would be used at compile time - and hope that's the same as was actually used when that particular Python package was installed. That's exactly what you've been trying to do so far. Which puts me back to my question: why not just replace this by _findLib_gcc(name)? It would not have any issues with files where the `SONAME` in the header is different from that in the filename, and it also wouldn't have issues if the `SONAME` is missing from the header alltogehter.

Reply:
I think I agree with your assessment, but it is quite a departure from what is currently happening (but what is happening is quite broken, for us at least)